### ISSUE TITLE
Use SoftFailure in Kinder

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#8506]
As part of getting rid of Validation in Kinder, we should start by refactoring it to use SoftFailure.

I think the plan is that every time we have something like:

```
case (args, t) => unify(t.kind, expectedKind) match {
  case Some(_) => Validation.success(Type.Alias(cst, args, t, loc))
  case None => Validation.toHardFailure(KindError.UnexpectedKind(expectedKind = expectedKind, actualKind = t.kind, loc))
}
```

We will replace it with a SoftFailure where we use Type.freshError with a Kind.Error.

**Comment by magnus-madsen** [comment]
@jaschdoc Does it make sense?

**Comment by magnus-madsen** [comment]
In KindEnv, I think if there is a conflict then we should update the map with Kind.Error.

**Comment by jaschdoc** [comment]
Sounds good

**Comment by jaschdoc** [comment]
Hmm I'm not sure it's as easy as that. The current errors are unrecoverable, and the soft failure requires recoverable errors. Making them recoverable is dangerous unless we handle that in Typer or somewhere else.

**Comment by magnus-madsen** [comment]
Is replacing types with Type.Error not enough?

**Comment by jaschdoc** [comment]
Not quite. There were a few later functions that assumed the `Kinder` would stop compilation (so some cases would be unreachable / throw an error) and `visitTypeVarSym` uses the `UnexpectedKind` error at a place that should return `Validation[Symbol, KindError]` where we cannot return `Typpe.Error`. Currently, I have just return the symbol with a `Kind.Error` in it. See https://github.com/flix/flix/pull/8520/files

**Comment by magnus-madsen** [comment]
> Not quite. There were a few later functions that assumed the `Kinder` would stop compilation (so some cases would be unreachable / throw an error) and `visitTypeVarSym` uses the `UnexpectedKind` error at a place that should return `Validation[Symbol, KindError]` where we cannot return `Typpe.Error`. Currently, I have just return the symbol with a `Kind.Error` in it. See https://github.com/flix/flix/pull/8520/files

I am sorry, I did not understand any of that. 

Are you saying we are blocked or that Type.Error works but you need extra tricks?

**Comment by magnus-madsen** [comment]
On a second read, I think it makes sense.

Does it seem to work fine?

Can you fire up VSCode, add some kind errors, and see if everything still works OK or you get compiler exceptions?

**Comment by magnus-madsen** [comment]
Can you open ticket with that crash? Thanks
